File: Assets/Scripts/Mechanics/ScenesManager.js
#pragma strict
var newScene:UnityEngine.SceneManagement.SceneManager; //Variável para a troca de cenas
static var instanceRef : ScenesManager; //Instancia do próprio script (que possui o nome de ScenesManager)

var duration : float; //Variavel para fazer contagem
var newGame : boolean; // Booleana para troca de cena
var loadGame : boolean; // Booleana para a troca de cena

static var callFadeInAgain : boolean; //CHAMAR O FADE IN (SERVE APENAS PARA A CENA "MENU")

// ################ VARIÁVEIS QUE CONTROLAM A ENTRADA DE CENAS ##########################

static var sceneMenu : boolean;
static var sceneSavenLoad : boolean;
static var sceneWorldMap : boolean;
static var scenePhase1 : boolean;


function Start () { //Começa chamando o fadeIn e setando algumas variáveis

	FadeInOutBehaviour.playfadeIn = true;

	duration = 0;
	newGame = false;
	loadGame = false;

		if(instanceRef == null){ //Também verifica se o objeto deve ser mantido
		DontDestroyOnLoad(this);
		}
}

function Update () {

	PermanentInGame(); //Verifica se a cena deve permanecer em game;

	CheckButtonWasPress(); // Realiza a ação de acordo com o botão que foi pressionado

	CallFadeInd(); // Chama o FadeIn quando necessário

	checkFunctions(); // Verifica quais funções precisam ser chamadas
 	
}

/* ######################## FUNÇÕES PARA EXECUÇÃO DO SCRIPT ############################
########################### AS FUNÇÕES DESSE PONTO EM DIANTE SÃO RELACIONADAS A: #######
########################### CHAMAR FADEIND/OUT #########################################
########################### VERIFICAR QUAL BOTÃO FOI PRESSIONADO #######################
########################### GARANTIR A PERMANENCIA DO SCRIPT DURANTE A TROCA DE CENAS ## 
########################### VERIFICAR A NECESSIDADE DE CHAMAR ALGUM MÉTODO #############*/

function checkFunctions(){ // Verifica quais funções precisam ser chamadas

		if(sceneMenu == true){
			sceneMenu = false;
			callMenu();
		}
		if(sceneSavenLoad == true){
			sceneSavenLoad = false;
			callSavenLoad();
		}
		if(sceneWorldMap == true){
			sceneWorldMap = false;
			callWorldMap();
		}
		if(scenePhase1 == true){
			scenePhase1 = false;
			callPhase1();
		}

}

function CallFadeInd(){
	if(callFadeInAgain == true){
	 		FadeInOutBehaviour.playfadeIn = true;
	 		callFadeInAgain = false;
	 	}
}

function CheckButtonWasPress(){

if( newGame == true){
		 		if(duration >= 1){
		 			newGame = false;
		 			dataBehaviour.newgameOption = true;
		 			duration = 0;
		 			callSavenLoad();
		 		} else {
		 			duration += Time.deltaTime;
		 		}
 	}
 	if( loadGame == true){
		 		if(duration >= 1){
		 			loadGame = false;
		 			dataBehaviour.loadingOption = true;
		 			duration = 0;
		 			callSavenLoad();
		 		} else {
		 			duration += Time.deltaTime;
		 		}
 	}

}

function PermanentInGame(){
if(newScene.GetActiveScene().name == "SavenLoad"){
		instanceRef = this;
		}

		if(newScene.GetActiveScene().name == "Menu" && instanceRef != null){
		callFadeInAgain = true;
		instanceRef = null;
		Destroy(this.gameObject);
		}
		if(instanceRef == null){
		DontDestroyOnLoad(this);
		}
}

/* ############## FUNÇÕES QUE ESTÃO PROGRAMADAS NOS BOTÕES DA TELA MENU ##########
################# ESSAS SÃO AS FUNÇÕES QUE FUNCIONAM APENAS NA CENA "MENU"########
################# TODAS ESTÃO SETADAS EM BOTÕES ################################## */


function NewGame(){
	newGame = true;
	dataBehaviour.callFadeInAgain = true;
}
function LoadGame(){
	loadGame = true;
	dataBehaviour.callFadeInAgain = true;
}

function fadeOut(){
	FadeInOutBehaviour.playfadeOut = true;
}

function QuitGame(){
				if(duration >= 1){
				Application.Quit();
				duration = 0;
			 		} else {
		 			duration += Time.deltaTime;
		 		}
}

/* ---------------- CONTROLE DA TRANSIÇÃO DE CENAS -----------------------*/

function callWorldMap(){
	newScene.LoadScene("worldmap");
}
function callPhase1(){
	newScene.LoadScene("phase_1");
}
function callMenu(){
	newScene.LoadScene("Menu");
}
function callSavenLoad(){
	newScene.LoadScene("SavenLoad");
}

File: Assets/Scripts/Mechanics/scenes_manager.cs
using UnityEngine;
using System.Collections;
using UnityEngine.SceneManagement;

public class scenes_manager : MonoBehaviour
{
    static scenes_manager instanceRef;

    public bool newGame;
    public bool loadGame;

    public bool sceneMenu;
    public bool sceneGamePlay;

    fade_behaviour FadeInOutBehaviour;
    data_savenload dataSaveNLoad;

    public Scene activeScene;

    void Start()
    {
        loadCurrentFade();
        permanentInGame();

        newGame = false;
        loadGame = false;
    }

    void Update()
    {
        sceneAction();
        sceneControl();
    }

    void loadCurrentFade()
    {
        FadeInOutBehaviour = GameObject.Find("fade_in_out").GetComponent<fade_behaviour>();
    }

    void callFadeOut()
    {
        FadeInOutBehaviour.playfadeOut = true;
    }

    void callFadeIn()
    {
        FadeInOutBehaviour.playfadeIn = true;
    }

    public void sceneControl()
    {
        if (activeScene != SceneManager.GetActiveScene())
        {
            activeScene = SceneManager.GetActiveScene();
            loadCurrentFade();
            switch (activeScene.name)
            {
                case "Menu":
                    sceneMenu = false;
                    callFadeIn();
                    break;
                case "SavenLoad":
                    dataSaveNLoad = GameObject.Find("UI").GetComponent<data_savenload>();
                    if (newGame == true)
                    {
                        dataSaveNLoad.newgameOption = true;
                        newGame = false;
                    }
                    if (loadGame == true)
                    {
                        dataSaveNLoad.loadingOption = true;
                        loadGame = false;
                    }
                    callFadeIn();
                    break;
                case "Gameplay":
                    sceneGamePlay = false;
                    callFadeIn();
                    break;
            }
        }
    }

    void sceneAction()
    {
        if (sceneMenu == true)
        {
            callMenu();
            instanceRef = null;
            Destroy(this.gameObject);
        }
        if (sceneGamePlay == true)
        {
            callGamePlay();
        }
        if (newGame == true || loadGame == true)
        {
            callSavenLoad();
        }
    }

    void permanentInGame()
    {
        if (instanceRef == null)
        {
            DontDestroyOnLoad(this);
            instanceRef = this;
        }
    }

    void callMenu()
    {
        SceneManager.LoadScene("Menu");
    }
    void callGamePlay()
    {
        SceneManager.LoadScene("Gameplay");
    }
    void callSavenLoad()
    {
        SceneManager.LoadScene("SavenLoad");
    }
}